modules/core/src/mathfuncs.cpp

```diff
@@ -1600,50 +1608,311 @@ static bool ocl_patchNaNs( InputOutputArray _a, float value )
 void patchNaNs( InputOutputArray _a, double _val )
 {
     CV_INSTRUMENT_REGION();
-
-    CV_Assert( _a.depth() == CV_32F );
+    CV_Assert( _a.depth() == CV_32F || _a.depth() == CV_64F);
 
     CV_OCL_RUN(_a.isUMat() && _a.dims() <= 2,
-               ocl_patchNaNs(_a, (float)_val))
+               ocl_patchNaNs(_a, _val))
 
     Mat a = _a.getMat();
     const Mat* arrays[] = {&a, 0};
     int* ptrs[1] = {};
     NAryMatIterator it(arrays, (uchar**)ptrs);
     size_t len = it.size*a.channels();
-    Cv32suf val;
-    val.f = (float)_val;
+
+    if (_a.depth() == CV_32F)
+    {
+        Cv32suf val;
+        val.f = (float)_val;
 
 #if CV_SIMD
-    v_int32 v_mask1 = vx_setall_s32(0x7fffffff), v_mask2 = vx_setall_s32(0x7f800000);
-    v_int32 v_val = vx_setall_s32(val.i);
+        v_int32 v_mask1 = vx_setall_s32(0x7fffffff), v_mask2 = vx_setall_s32(0x7f800000);
+        v_int32 v_val = vx_setall_s32(val.i);
 #endif
 
-    for( size_t i = 0; i < it.nplanes; i++, ++it )
+        for (size_t i = 0; i < it.nplanes; i++, ++it)
+        {
+            int *tptr = ptrs[0];
+            size_t j = 0;
+
+#if CV_SIMD
+            size_t cWidth = (size_t)v_int32::nlanes;
+            for (; j + cWidth <= len; j += cWidth)
+            {
+                v_int32 v_src = vx_load(tptr + j);
+                v_int32 v_cmp_mask = v_mask2 < (v_src & v_mask1);
+                v_int32 v_dst = v_select(v_cmp_mask, v_val, v_src);
+                v_store(tptr + j, v_dst);
+            }
+            vx_cleanup();
+#endif
+
+            for (; j < len; j++)
+                if ((tptr[j] & 0x7fffffff) > 0x7f800000)
+                    tptr[j] = val.i;
+        }
+    }
+    else if (_a.depth() == CV_64F)
     {
-        int* tptr = ptrs[0];
-        size_t j = 0;
+        Cv64suf val;
+        val.f = _val;
 
 #if CV_SIMD
-        size_t cWidth = (size_t)v_int32::nlanes;
-        for ( ; j + cWidth <= len; j += cWidth)
+        v_int64 v_mnt_mask = vx_setall_s64(0x000FFFFFFFFFFFFF), v_exp_mask = vx_setall_s64(0x7FF0000000000000);
+        v_int64 v_val = vx_setall_s64(val.i);
+#if !CV_SIMD128_64F
+        v_int64 mask10 = vx_setall_s64(0xffffffff00000000);
```

Comment:
> `CV_SIMD128_64F`

`64F` is about `double` (`float64`) type.
Using it to limit `int64` processing is wrong.

Reply:
Sure, but this is how to enable int64 comparison in NEON universal intrinsics: [intrin_neon.hpp](https://github.com/opencv/opencv/blob/23dec329b47cf19ce2bbbb623c1467929ac021e3/modules/core/include/opencv2/core/hal/intrin_neon.hpp#L1073)

Reply:
Currently comparison of 64-bit integer SIMD is **declared as non-supported**:

https://github.com/opencv/opencv/blame/4.7.0/modules/core/include/opencv2/core/hal/intrin_cpp.hpp#L885

```
For all types except 64-bit integer values.
```

No idea why NEON hijacks that and provides some implementation (only for `v_uint64x2`, but not for signed `v_int64x2`).
Probably added by mistake here: #7175 (patch should target 64F only).
Also there is contributed test for eq/ne 64-bit here: #15738 (with discussion of misused macro)

---

Perhaps we need to allow and implement this support for eq/ne (`==`/`!=`) comparisons at least for all SIMD backends.

/cc @mshabunin @vpisarev

Reply:
Since this works now w/o workarounds, I've rewritten it in a more convenient way